### DISCUSSION TITLE
On Error Resume Next

### DISCUSSION THREAD

**Discussion by cristianbuse** [Theadd/ds3x#28] [General]
Hi @Theadd ,

There is an annoying inconsistency with error handling in VBA. When using an ```On Error GoTo Label``` the global ```Err``` object will get reset before exiting the method if an error occurs. Not for ```On Error Resume Next``` though. If an error occurs then the ```Err``` object retains the error information even after exiting the scope of the method unless it gets reset. To make sure this does not happen and logic like ```If Err.Number = 0 Then``` is not affected up the calling chain, we need to reset the error with ```On Error GoTo 0```.

Here is proof:
```VBA
Sub TestErr()
    ErrGo
    Debug.Print "Error is " & Err.Number & " when using On Error GoTo"
    
    ErrResume
    Debug.Print "Error is " & Err.Number & " when using On Error Resume Next"
End Sub

Sub ErrGo()
    Dim i As Long
    On Error GoTo ErrHandler
    i = 1 / 0
Exit Sub
ErrHandler:
    'Do nothing
End Sub

Sub ErrResume()
    Dim i As Long
    On Error Resume Next
    i = 1 / 0
End Sub
```
Running ```TestErr``` will display:
![image](https://github.com/Theadd/ds3x/assets/23198997/17396c20-2d76-44cf-bb6b-7978cbc413a6)

So, the error is propagated outside of the routine which is quite bad if you have code that relies on the ```Err``` object to be reset properly.

Always use a closing ```On Error GoTo 0``` after you are done. For example [this](https://github.com/Theadd/ds3x/blob/ae76a6b3d37afd8bce698145d5af53e28a05b194/src/ds3x.accdb.src/modules/Array2dEx.cls#L23-L26) should become:
```VBA
Public Property Get Count() As Long
    On Error Resume Next
    Count = 1 + (UBound(Instance, 1) - LBound(Instance, 1))
    On Error GoTo 0 'Or Err.Clear
End Property
```

**Comment by Theadd** [comment]
Hi @cristianbuse and thanks for all the feedback!

That error propagation problem is nice to know indeed, it made me remember myself a few times with a "WTF" expression when printing the `Err.Description`, now I understand why it didn't make any sense.

https://github.com/Theadd/ds3x/commit/53d3653d1ff9156abf65e2ba26ffc38dfd6ebc13 I fixed them all.

Thank you!
